ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateMutators.java

```diff
@@ -98,18 +110,91 @@ public void initiateValidatorExit(MutableBeaconState state, int index) {
         .set(
             index,
             validator
-                .withExitEpoch(exitQueueEpoch)
+                .withExitEpoch(validatorExitContext.exitQueueEpoch)
                 .withWithdrawableEpoch(
-                    exitQueueEpoch.plus(specConfig.getMinValidatorWithdrawabilityDelay())));
+                    validatorExitContext.exitQueueEpoch.plus(
+                        specConfig.getMinValidatorWithdrawabilityDelay())));
+  }
+
+  /**
+   * We want the context to be lazily initialized, so we use a memoizing supplier.
+   *
+   * @param state
+   * @return
+   */
+  public Supplier<ValidatorExitContext> createValidatorExitContextSupplier(
+      final BeaconState state) {
+    return Suppliers.memoize(() -> createValidatorExitContext(state));
+  }
+
+  /**
+   * This function implements an optimized version of exitQueueEpoch and exitQueueChurn calculation,
+   * compared to the `initiateValidatorExit` reference implementation.
```

Comment:
I would probably use the python function name when talking about reference implementation...